INSP: It is not allowed to cast to a bool. (E0054)

zeroeightysix opened the pull request:
This PR aims to implement an inspection for rust error [E0054](https://doc.rust-lang.org/error-index.html#E0054) *It is not allowed to cast to a bool.*

If this is merged, I am also interested in implementing an intent to "compare with zero instead", as the error index (and compiler) suggests.

Part of #886

zeroeightysix commented on src/main/resources/META-INF/rust-core.xml:
```diff
@@ -524,6 +524,11 @@
                          enabledByDefault="true" level="ERROR"
                          implementationClass="org.rust.ide.inspections.RsTraitImplementationInspection"/>
 
+        <localInspection language="Rust" groupName="Rust"
+                         displayName="Cast to bool"
+                         enabledByDefault="true" level="ERROR"
+                         implementationClass="org.rust.ide.inspections.RsCastToBoolInspection"/>
+
         <localInspection language="Rust" groupPath="Rust" groupName="Lints"
```
Is there an ordering to be followed in this file? I tossed the entry next to the intent for E0050 (as this PR is for E0054, so they line up).

zeroeightysix commented on src/test/kotlin/org/rust/ide/inspections/RsCastToBoolInspectionTest.kt:
```diff
@@ -0,0 +1,31 @@
+        }
+    """)
+
+    fun testCastBoolToBool() = checkByText("""
+        fn main() {
+            true as bool;
+        }
+    """)
```
I wasn't sure how else to test this, but what is being tested here is that there *aren't* any errors in the provided code. I'd love to know if there's a better way to do it!

Undin replied:
It's exactly how we test absence of errors. If `RsCastToBoolInspection` start produce error in this case, test will fail

Undin commented on src/test/kotlin/org/rust/ide/inspections/RsCastToBoolInspectionTest.kt:
```diff
@@ -0,0 +1,31 @@
+
+class RsCastToBoolInspectionTest : RsInspectionsTestBase(RsCastToBoolInspection::class) {
+
+    fun testCastNumToBool() = checkByText("""
```
```suggestion
    fun `test cast num to bool`() = checkByText("""
```
We use names with spaces enclosed in backticks for test methods. [See](https://kotlinlang.org/docs/reference/coding-conventions.html#names-for-test-methods)

zeroeightysix replied:
Oops. I used #713 as reference (where this style was not used), as it was listed in the contributing page [here](https://github.com/intellij-rust/intellij-rust/blob/master/CONTRIBUTING.md#contributing). Might want to update those to more recent PRs, to prevent cases like these.

Undin commented on src/main/kotlin/org/rust/ide/inspections/RsCastToBoolInspection.kt:
```diff
@@ -0,0 +1,32 @@
+            // It is allowed to cast a bool to a bool, so if the expression's type is of bool, ignore this cast.
+            // Casts from non primitive types (and the unit type) emit another error, so we ignore those as well.
+            val exprType = castExpr.expr.type
+            if (exprType === TyBool || exprType !is TyPrimitive || exprType is TyUnit) return
```
```suggestion
            if (exprType === TyBool || exprType !is TyPrimitive || exprType == TyUnit) return
```

Undin reviewed (commented):
LGTM
Could you squash "fixing" commits in main ones?

zeroeightysix commented:
> Could you squash "fixing" commits in main ones?

Sure, but I'm not sure what you mean by 'main ones'. Do you want me to group commits *by tag*, so we get something like
```
INSP: Added cast to bool inspection.
T: Added tests for cast to bool inspection
DOC: Added zeroeightysix to CONTRIBUTORS.md
```

Or squash all into one commit (with or without (multiple) tags)?

Undin commented:
@zeroeightysix we don't have strict rules about number/structure of commits. We just prefer not to have separate commits with review fixes in master because they have a little sense there. Other decisions about commit history are up to you. 

> INSP: Added cast to bool inspection.
> T: Added tests for cast to bool inspection
> DOC: Added zeroeightysix to CONTRIBUTORS.txt

is ok for me

zeroeightysix commented:
All done :+1:

Undin reviewed (approved):
Thank you!
Welcome to IntelliJ Rust contributors!

bors r+

Undin commented:
bors retry
